src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -676,14 +726,22 @@ public void RemoveDynamicSourceFile(string dynamicFilePath)
 
         private void OnDynamicFileInfoUpdated(object sender, string dynamicFilePath)
         {
-            if (!_dynamicFilePathMaps.TryGetValue(dynamicFilePath, out var fileInfoPath))
+            string? fileInfoPath;
+
+            lock (_gate)
             {
-                // given file doesn't belong to this project. 
-                // this happen since the event this is handling is shared between all projects
-                return;
+                if (!_dynamicFilePathMaps.TryGetValue(dynamicFilePath, out fileInfoPath))
```

Comment:
did you audit all usages of _dynamicFilePathMaps/_sourceFiles?  if so, this PR LGTM.

Reply:
Yes, all uses of _dynamicFilePathMaps is audited. _sourceFiles itself takes the lock so it's not required to be in the lock, but we're good.